cdap-security/src/main/java/io/cdap/cdap/security/auth/KeyIdentifierCodec.java

```diff
@@ -67,11 +87,21 @@ public KeyIdentifier decode(byte[] data) throws IOException {
 
     DatumReader<KeyIdentifier> reader = readerFactory.create(KEY_IDENTIFIER_TYPE,
                                                              KeyIdentifier.Schemas.getCurrentSchema());
-    int readVersion = decoder.readInt();
-    Schema readSchema = KeyIdentifier.Schemas.getSchemaVersion(readVersion);
-    if (readSchema == null) {
-      throw new IOException("Unknown schema version for KeyIdentifier: " + readVersion);
+    try {
+      int readVersion = decoder.readInt();
+      Schema readSchema = KeyIdentifier.Schemas.getSchemaVersion(readVersion);
+      if (readSchema == null) {
+        throw new IOException("Unknown schema version for KeyIdentifier: " + readVersion);
+      }
+      return reader.read(decoder, readSchema);
+    } catch (IOException e) {
+      // As a fallback, try deserializing from JSON format.
+      try {
+        return gson.fromJson(new String(data, StandardCharsets.UTF_8), KeyIdentifier.class);
```

Comment:
What would happen if the original data is not a valid UTF_8 string? Overall, what is the use case for this fallback?

Reply:
From the javadoc on the String constructor:
```
This method always replaces malformed-input and unmappable-character sequences with this charset's default replacement string.
```

As such, I expect it will decode to garbage and then throw a JsonSyntaxException. This JSON fallback is necessary so that we can create keys in other languages without having to port the serialization logic over to the other language.